File: apps/android/app/src/main/java/com/ameme/android/data/transport/AgentAccessGrant.kt
package com.ameme.android.data.transport

import java.time.Instant

/**
 * Android representation of the canonical AccessGrant contract.
 *
 * This is authorization metadata only: it contains no pairing secret, endpoint,
 * certificate private key, or memory content. A channel must authenticate the
 * peer separately and then authorize each minimal request against this object.
 */
data class AgentAccessGrant(
    val schemaVersion: Int,
    val grantId: String,
    val ownerId: String,
    val callerId: String,
    val purposes: Set<String>,
    val spaces: Set<String>,
    val dataTypes: Set<String>,
    val notBefore: Instant,
    val expiresAt: Instant,
    val status: AgentAccessGrantStatus,
    val keyFingerprint: String? = null,
    val createdAt: Instant,
    val revokedAt: Instant? = null,
) {
    init {
        require(schemaVersion == CURRENT_SCHEMA_VERSION) { "unsupported AccessGrant schema" }
        requireIdentifier(grantId, "grantId")
        requireIdentifier(ownerId, "ownerId")
        requireIdentifier(callerId, "callerId")
        requireScope(purposes, "purposes", MAX_PURPOSE_LENGTH)
        requireScope(spaces, "spaces", MAX_IDENTIFIER_LENGTH)
        requireScope(dataTypes, "dataTypes", MAX_DATA_TYPE_LENGTH)
        require(!expiresAt.isBefore(notBefore)) { "grant expiry must not precede notBefore" }
        require(expiresAt.isAfter(createdAt)) { "grant expiry must be in the future of creation" }
        keyFingerprint?.let { require(it.length <= MAX_KEY_FINGERPRINT_LENGTH) }
        if (status == AgentAccessGrantStatus.Revoked) {
            require(revokedAt != null) { "revoked grant requires revokedAt" }
        } else {
            require(revokedAt == null) { "only revoked grant may have revokedAt" }
        }
    }

    fun isActive(at: Instant = Instant.now()): Boolean =
        status == AgentAccessGrantStatus.Active &&
            !at.isBefore(notBefore) &&
            expiresAt.isAfter(at)

    /** Checks a minimal request scope; a request may never expand this Grant. */
    fun authorize(request: AgentAccessGrantRequest, at: Instant = Instant.now()) {
        authorizeScope(
            callerId = request.callerId,
            grantId = request.grantId,
            purpose = request.purpose,
            spaces = setOf(request.space),
            dataTypes = setOf(request.dataType),
            at = at,
        )
    }

    fun authorizeScope(
        callerId: String,
        grantId: String,
        purpose: String,
        spaces: Set<String>,
        dataTypes: Set<String>,
        at: Instant = Instant.now(),
    ) {
        if (callerId != this.callerId || grantId != this.grantId) {
            throw AgentAccessGrantException(AgentAccessGrantFailure.AuthRequired)
        }
        when {
            status == AgentAccessGrantStatus.Revoked ->
                throw AgentAccessGrantException(AgentAccessGrantFailure.GrantRevoked)
            !expiresAt.isAfter(at) || at.isBefore(notBefore) ->
                throw AgentAccessGrantException(AgentAccessGrantFailure.GrantExpired)
            purpose !in purposes ->
                throw AgentAccessGrantException(AgentAccessGrantFailure.PurposeDenied)
            !spaces.all(this.spaces::contains) ->
                throw AgentAccessGrantException(AgentAccessGrantFailure.SpaceDenied)
            !dataTypes.all(this.dataTypes::contains) ->
                throw AgentAccessGrantException(AgentAccessGrantFailure.DataTypeDenied)
        }
    }

    companion object {
        const val CURRENT_SCHEMA_VERSION = 1
        private const val MAX_SCOPE_ITEMS = 8
        private const val MAX_IDENTIFIER_LENGTH = 128
        private const val MAX_PURPOSE_LENGTH = 80
        private const val MAX_DATA_TYPE_LENGTH = 64
        private const val MAX_KEY_FINGERPRINT_LENGTH = 256
        private val IDENTIFIER = Regex("[A-Za-z0-9][A-Za-z0-9._:-]{0,127}")

        private fun requireIdentifier(value: String, field: String) {
            require(IDENTIFIER.matches(value)) { "$field is invalid" }
            require('*' !in value) { "$field must not contain a wildcard" }
        }

        private fun requireScope(values: Set<String>, field: String, maxLength: Int) {
            require(values.isNotEmpty() && values.size <= MAX_SCOPE_ITEMS) {
                "$field must contain a bounded non-empty scope"
            }
            require(values.all { it.isNotBlank() && it.length <= maxLength && '*' !in it }) {
                "$field contains an invalid wildcard or value"
            }
        }
    }
}

/**
 * Local, user-approved scope bound to a pairing. The caller and Grant IDs are
 * still bound by [AgentAccessGrant] when a request arrives; this policy only
 * records the maximum scope that the user approved for the pairing.
 */
data class AgentAccessGrantPolicy(
    val schemaVersion: Int,
    val ownerId: String,
    val purposes: Set<String>,
    val spaces: Set<String>,
    val dataTypes: Set<String>,
    val notBefore: Instant,
    val expiresAt: Instant,
    val status: AgentAccessGrantStatus,
    val createdAt: Instant,
) {
    init {
        require(schemaVersion == AgentAccessGrant.CURRENT_SCHEMA_VERSION)
        require(ownerId.matches(IDENTIFIER))
        require(purposes.isNotEmpty() && purposes.size <= 8 && purposes.all { it.length <= 80 && '*' !in it })
        require(spaces.isNotEmpty() && spaces.size <= 8 && spaces.all { it.length <= 128 && '*' !in it })
        require(dataTypes.isNotEmpty() && dataTypes.size <= 8 && dataTypes.all { it.length <= 64 && '*' !in it })
        require(expiresAt >= notBefore && expiresAt > createdAt)
    }

    fun isActive(at: Instant = Instant.now()): Boolean =
        status == AgentAccessGrantStatus.Active && at >= notBefore && at < expiresAt

    fun bind(callerId: String, grantId: String): AgentAccessGrant = AgentAccessGrant(
        schemaVersion = schemaVersion,
        grantId = grantId,
        ownerId = ownerId,
        callerId = callerId,
        purposes = purposes,
        spaces = spaces,
        dataTypes = dataTypes,
        notBefore = notBefore,
        expiresAt = expiresAt,
        status = status,
        createdAt = createdAt,
    )

    companion object {
        private val IDENTIFIER = Regex("[A-Za-z0-9][A-Za-z0-9._:-]{0,127}")

        fun default(
            ownerId: String,
            createdAt: Instant,
            expiresAt: Instant,
        ): AgentAccessGrantPolicy = AgentAccessGrantPolicy(
            schemaVersion = AgentAccessGrant.CURRENT_SCHEMA_VERSION,
            ownerId = ownerId,
            purposes = setOf("autonomous_memory"),
            spaces = setOf("space_personal"),
            dataTypes = setOf("event"),
            notBefore = createdAt,
            expiresAt = expiresAt,
            status = AgentAccessGrantStatus.Active,
            createdAt = createdAt,
        )
    }
}

enum class AgentAccessGrantStatus(val wireValue: String) {
    Active("active"),
    Revoked("revoked"),
    Expired("expired"),
}

data class AgentAccessGrantRequest(
    val callerId: String,
    val grantId: String,
    val purpose: String,
    val space: String,
    val dataType: String,
    val operation: String,
) {
    init {
        require(callerId.matches(IDENTIFIER))
        require(grantId.matches(IDENTIFIER))
        require(purpose.isNotBlank() && purpose.length <= 80)
        require(space.isNotBlank() && space.length <= 128)
        require(dataType.isNotBlank() && dataType.length <= 64)
        require(operation.matches(IDENTIFIER))
    }

    private companion object {
        val IDENTIFIER = Regex("[A-Za-z0-9][A-Za-z0-9._:-]{0,127}")
    }
}

enum class AgentAccessGrantFailure(val wireValue: String) {
    AuthRequired("AUTH_REQUIRED"),
    GrantRevoked("GRANT_REVOKED"),
    GrantExpired("GRANT_EXPIRED"),
    PurposeDenied("PURPOSE_DENIED"),
    SpaceDenied("SPACE_DENIED"),
    DataTypeDenied("DATA_TYPE_DENIED"),
}

class AgentAccessGrantException(
    val failure: AgentAccessGrantFailure,
) : IllegalStateException(failure.wireValue)

File: apps/ios/Ameme/Shared/AgentAccessGrant.swift
import Foundation

/// Swift representation of the canonical AccessGrant contract.
///
/// This is authorization metadata only. Pairing secrets, endpoints, private
/// keys and memory content remain outside the Grant and its persistence layer.
public struct AgentAccessGrant: Codable, Equatable, Sendable {
    public static let currentSchemaVersion = 1

    public let schemaVersion: Int
    public let grantID: String
    public let ownerID: String
    public let callerID: String
    public let purposes: [String]
    public let spaces: [String]
    public let dataTypes: [String]
    public let notBefore: Date
    public let expiresAt: Date
    public let status: AgentAccessGrantStatus
    public let keyFingerprint: String?
    public let createdAt: Date
    public let revokedAt: Date?

    public init(
        grantID: String,
        ownerID: String,
        callerID: String,
        purposes: [String],
        spaces: [String],
        dataTypes: [String],
        notBefore: Date,
        expiresAt: Date,
        status: AgentAccessGrantStatus,
        keyFingerprint: String? = nil,
        createdAt: Date,
        revokedAt: Date? = nil
    ) {
        self.schemaVersion = Self.currentSchemaVersion
        self.grantID = grantID
        self.ownerID = ownerID
        self.callerID = callerID
        self.purposes = purposes
        self.spaces = spaces
        self.dataTypes = dataTypes
        self.notBefore = notBefore
        self.expiresAt = expiresAt
        self.status = status
        self.keyFingerprint = keyFingerprint
        self.createdAt = createdAt
        self.revokedAt = revokedAt
        precondition(Self.isValidIdentifier(grantID))
        precondition(Self.isValidIdentifier(ownerID))
        precondition(Self.isValidIdentifier(callerID))
        Self.preconditionScope(purposes, maxLength: 80)
        Self.preconditionScope(spaces, maxLength: 128)
        Self.preconditionScope(dataTypes, maxLength: 64)
        precondition(expiresAt >= notBefore)
        precondition(expiresAt > createdAt)
        precondition(keyFingerprint?.count ?? 0 <= 256)
        if status == .revoked {
            precondition(revokedAt != nil)
        } else {
            precondition(revokedAt == nil)
        }
    }

    public func isActive(at date: Date = .now) -> Bool {
        status == .active && date >= notBefore && date < expiresAt
    }

    public func authorize(_ request: AgentAccessGrantRequest, at date: Date = .now) throws {
        try authorizeScope(
            callerID: request.callerID,
            grantID: request.grantID,
            purpose: request.purpose,
            spaces: [request.space],
            dataTypes: [request.dataType],
            at: date
        )
    }

    /// Checks an explicit minimal request scope against this Grant.
    ///
    /// This mirrors the Android Local Node authorization boundary for bounded
    /// reads, whose payload may name more than one already-granted space/type.
    public func authorizeScope(
        callerID: String,
        grantID: String,
        purpose: String,
        spaces: Set<String>,
        dataTypes: Set<String>,
        at date: Date = .now
    ) throws {
        guard callerID == self.callerID, grantID == self.grantID else {
            throw AgentAccessGrantError.authRequired
        }
        switch status {
        case .revoked:
            throw AgentAccessGrantError.grantRevoked
        case .expired:
            throw AgentAccessGrantError.grantExpired
        case .active:
            guard date >= notBefore, date < expiresAt else {
                throw AgentAccessGrantError.grantExpired
            }
        }
        guard purposes.contains(purpose) else {
            throw AgentAccessGrantError.purposeDenied
        }
        guard !spaces.isEmpty, spaces.isSubset(of: Set(self.spaces)) else {
            throw AgentAccessGrantError.spaceDenied
        }
        guard !dataTypes.isEmpty, dataTypes.isSubset(of: Set(self.dataTypes)) else {
            throw AgentAccessGrantError.dataTypeDenied
        }
    }

    private static func isValidIdentifier(_ value: String) -> Bool {
        value.range(of: "^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$", options: .regularExpression) != nil
    }

    private static func preconditionScope(_ values: [String], maxLength: Int) {
        precondition(!values.isEmpty && values.count <= 8)
        precondition(Set(values).count == values.count)
        precondition(values.allSatisfy { !$0.isEmpty && $0.count <= maxLength && !$0.contains("*") })
    }
}

/// The maximum scope the user approved locally for one pairing.
///
/// The host still supplies the caller and Grant identifiers. Binding those
/// identifiers to this policy creates the request-time [AgentAccessGrant]
/// checked by the local node; the policy itself never contains channel secrets.
public struct AgentAccessGrantPolicy: Codable, Equatable, Sendable {
    public let schemaVersion: Int
    public let ownerID: String
    public let purposes: [String]
    public let spaces: [String]
    public let dataTypes: [String]
    public let notBefore: Date
    public let expiresAt: Date
    public let status: AgentAccessGrantStatus
    public let createdAt: Date

    public init(
        ownerID: String,
        purposes: [String],
        spaces: [String],
        dataTypes: [String],
        notBefore: Date,
        expiresAt: Date,
        status: AgentAccessGrantStatus,
        createdAt: Date
    ) {
        self.schemaVersion = AgentAccessGrant.currentSchemaVersion
        self.ownerID = ownerID
        self.purposes = purposes
        self.spaces = spaces
        self.dataTypes = dataTypes
        self.notBefore = notBefore
        self.expiresAt = expiresAt
        self.status = status
        self.createdAt = createdAt
        precondition(Self.isValidIdentifier(ownerID))
        Self.preconditionScope(purposes, maxLength: 80)
        Self.preconditionScope(spaces, maxLength: 128)
        Self.preconditionScope(dataTypes, maxLength: 64)
        precondition(expiresAt >= notBefore && expiresAt > createdAt)
    }

    public func isActive(at date: Date = .now) -> Bool {
        status == .active && date >= notBefore && date < expiresAt
    }

    public func bind(callerID: String, grantID: String) -> AgentAccessGrant {
        AgentAccessGrant(
            grantID: grantID,
            ownerID: ownerID,
            callerID: callerID,
            purposes: purposes,
            spaces: spaces,
            dataTypes: dataTypes,
            notBefore: notBefore,
            expiresAt: expiresAt,
            status: status,
            createdAt: createdAt
        )
    }

    public static func `default`(createdAt: Date, expiresAt: Date) -> AgentAccessGrantPolicy {
        AgentAccessGrantPolicy(
            ownerID: "owner_local",
            purposes: ["autonomous_memory"],
            spaces: ["space_personal"],
            dataTypes: ["event"],
            notBefore: createdAt,
            expiresAt: expiresAt,
            status: .active,
            createdAt: createdAt
        )
    }

    private static func isValidIdentifier(_ value: String) -> Bool {
        value.range(of: "^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$", options: .regularExpression) != nil
    }

    private static func preconditionScope(_ values: [String], maxLength: Int) {
        precondition(!values.isEmpty && values.count <= 8)
        precondition(Set(values).count == values.count)
        precondition(values.allSatisfy { !$0.isEmpty && $0.count <= maxLength && !$0.contains("*") })
    }
}

public enum AgentAccessGrantStatus: String, Codable, Sendable {
    case active
    case revoked
    case expired
}

public struct AgentAccessGrantRequest: Codable, Equatable, Sendable {
    public let callerID: String
    public let grantID: String
    public let purpose: String
    public let space: String
    public let dataType: String
    public let operation: String

    public init(
        callerID: String,
        grantID: String,
        purpose: String,
        space: String,
        dataType: String,
        operation: String
    ) {
        self.callerID = callerID
        self.grantID = grantID
        self.purpose = purpose
        self.space = space
        self.dataType = dataType
        self.operation = operation
        precondition(callerID.range(of: "^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$", options: .regularExpression) != nil)
        precondition(grantID.range(of: "^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$", options: .regularExpression) != nil)
        precondition(!purpose.isEmpty && purpose.count <= 80)
        precondition(!space.isEmpty && space.count <= 128)
        precondition(!dataType.isEmpty && dataType.count <= 64)
        precondition(operation.range(of: "^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$", options: .regularExpression) != nil)
    }
}

public enum AgentAccessGrantError: Error, Equatable, Sendable {
    case authRequired
    case grantRevoked
    case grantExpired
    case purposeDenied
    case spaceDenied
    case dataTypeDenied
}
